app/controllers/ConfirmSubmissionController.scala

```diff
@@ -16,33 +16,49 @@
 
 package controllers
 
-import java.time.LocalDate
-
+import common.SessionKeys
 import config.{AppConfig, ErrorHandler}
 import controllers.predicates.{AuthPredicate, MandationStatusPredicate}
-import javax.inject.Inject
-import models._
+import javax.inject.{Inject, Singleton}
+import models.{ConfirmSubmissionViewModel, SubmitVatReturnModel}
 import play.api.i18n.{I18nSupport, MessagesApi}
 import play.api.libs.json.Json
 import play.api.mvc._
+import services.VatSubscriptionService
 import uk.gov.hmrc.play.bootstrap.controller.FrontendController
 
 import scala.concurrent.Future
 
+@Singleton
 class ConfirmSubmissionController @Inject()(val messagesApi: MessagesApi,
                                             val mandationStatusCheck: MandationStatusPredicate,
                                             val errorHandler: ErrorHandler,
+                                            val vatSubscriptionService: VatSubscriptionService,
                                             authPredicate: AuthPredicate,
                                             implicit val appConfig: AppConfig) extends FrontendController with I18nSupport {
 
-  def show(frs: Boolean, obs: String, nbm: String, name: Option[String], periodKey: String): Action[AnyContent] = Action.async { implicit request =>
-    val viewModel = ConfirmSubmissionViewModel(
-      Json.parse(obs).as[VatObligations].obligations.head, frs, Json.parse(nbm).as[NineBoxModel], name
-    )
-    Future.successful(Ok(views.html.confirm_submission(viewModel)))
+  def show(periodKey: String): Action[AnyContent] = (authPredicate andThen mandationStatusCheck).async { implicit user =>
+
+    user.session.get(SessionKeys.viewModel) match {
+      case Some(model) => {
+        val sessionData = Json.parse(model).as[SubmitVatReturnModel]
+
+        vatSubscriptionService.getCustomerDetails(user.vrn) map {
+          case (Right(customerDetails)) => {
+            val viewModel = ConfirmSubmissionViewModel(sessionData, periodKey, customerDetails.clientName)
+            Ok(views.html.confirm_submission(viewModel))
+          }
+          case _ => {
+            val viewModel = ConfirmSubmissionViewModel(sessionData, periodKey, None)
+            Ok(views.html.confirm_submission(viewModel))
+          }
+        }
+      }
+      case _ => Future.successful(Redirect(controllers.routes.SubmitFormController.show(periodKey)))
+    }
   }
 
-  def submit(periodKey: String): Action[AnyContent] = Action.async { implicit request =>
-    Future.successful(Redirect(controllers.routes.ConfirmationController.show().url))
+  def submit(periodKey: String): Action[AnyContent] = (authPredicate andThen mandationStatusCheck) { implicit user =>
+    Redirect(controllers.routes.ConfirmationController.show().url)
```

Comment:
Could you add removal of session data in this please? Then when we do the actual backend submission we don't forget to clear it after it being successful

Reply:
This is being done under another task